### DISCUSSION TITLE
How exactly indents work?

### DISCUSSION THREAD

**Discussion by MunifTanjim** [nvim-treesitter/nvim-treesitter#1668] [Q&A] [answered]
I tried to read the implementation code (`indent.lua`), but it's pretty hard for me to understand what's acutally going on.

Can somebody please explain how the `@indent`, `@branch` etc. nodes are used to do the indenting?

I'm asking cause I can't make indents for **lua** `if .. else` work as expected. If I type the following block of code:

```lua
if cond then
  print('if')
  else
    print('else')
end
```

the `else` block is always indented one level deeper than expected 🤔 

P.S.:  I'm happy to help in any way I can to make the indenting system better.

**Comment by vigoux** [comment]
Hi, I am not an expert in this, so just pinging the relevant people : @stsewd @kyazdani42

**Comment by kyazdani42** [comment, marked as answer]
hi !
the algorithm is quite wonky atm, but basically, `@indent` branches are supposed to denote a node to indent, `@branch` is supposed to dedent the node, and `@ignore` is supposed to skip the node.
I think if we could find a better way to do this it would be better, i wrote the first implementation of this algorithm, and it is still very incomplete i believe. Plus i don't have time to work on this at all :( I'd be happy to try implement something better or improve the existing algorithm but i'd need time, a lot of it :smile:

**Reply by kyazdani42** [reply]
the hard part is when dealing with new line indents, because most of the time, the node is not finished `if something { -- here node is erroring` and each language behave differently (take python vs c/rust/go for instance)

**Reply by MunifTanjim** [reply]
Thanks!

I dug around a little bit more and I think we're missing another piece of the puzzle, which is `indentkeys`. Vim decides when to call `indentexpr` based on `indentkeys`. In my case the `else` was not dedenting because Vim was not calling `indexexpr` when I typed `else`.

It works as expected if I adjusted the option like this: `vim.bo.indentkeys = vim.bo.indentkeys .. ',0=else'`

**Reply by kyazdani42** [reply]
i guess this is more dependant on user configuration. If it works in normal mode, it should work in insert mode. But it's hard to handle indentation when user is typing because the nodes could be erroring depending on when it tries to indent.
